src/NUnitEngine/nunit-agent/Program.cs

```diff
@@ -130,58 +122,65 @@ public static void Main(string[] args)
             log.Info("Initializing Services");
             engine.Initialize();
 
-            log.Info("Starting RemoteTestAgent");
-            Agent = new RemoteTestAgent(AgentId, AgencyUrl, engine.Services);
-
+            log.Info("Starting AgentServer");
             try
             {
-                if (Agent.Start())
-                    WaitForStop();
-                else
+                using (var server = AgentServer.Start(IPAddress.Loopback, engine.Services.GetService<ITestRunnerFactory>()))
                 {
-                    log.Error("Failed to start RemoteTestAgent");
-                    Environment.Exit(AgentExitCodes.FAILED_TO_START_REMOTE_AGENT);
+                    // Write to the console to let the parent process know which port Windows assigned
+                    Console.WriteLine("Listening on " + server.ListeningOn);
+                    log.Info("Listening on " + server.ListeningOn);
+
+                    Console.WriteLine("Type 'stop' to exit.");
+
+                    while (true)
+                    {
+                        var line = Console.ReadLine();
```

Comment:
You are using a TCP channel to send results back to the parent process, but you are using the captured STDIN and STDOUT to communicate too. For example the port number above and the stop command here. That seems to defeat your goal of allowing communications with remote  agents like mobile devices and emulators.

Am I misreading this? Is this just another way to stop? I see you have a stop command too.

Reply:
The way I have it here, starting and stopping of the TCP server itself is set up using standard streams, and test runs (and stopping of test runs) is done over TCP. I was thinking of these as two distinct layers that don't mix.